### PR TITLE
Switched TableBlock’s header controls to a field that requires user input

### LINKED ISSUE #5989: Feature request: Switch TableBlock’s header controls to a field that requires user input
### Is your proposal related to a problem?

Related: #5865, #5977. Wagtail’s [TableBlock](https://docs.wagtail.io/en/v2.8.1/reference/contrib/table_block.html#tableblock) has support for defining the first row, or column, as table headers. 

While it’s not mandatory for a table to have header cells, if the table isn’t meant for layout then using header cells is a usability and accessibility benefit (or requirement) – header cells are generally displayed differently (e.g. bold), and screen readers will use the header cell’s label when announcing items is the header for. I believe failing to mark up header cells as such is a failure of [WCAG2.1’s Success Criterion 1.3.1 (Info and Relationships)](https://www.w3.org/TR/WCAG20-TECHS/F91.html).

I think the current form makes it a bit too easy to miss those checkboxes and leave the table without headers.

### Describe the solution you'd like

I would suggest changing the header cells’ two checkboxes to a single "Header cells" dropdown, with the following options:

- [Choose] (Default)
- First row
- First column
- First row and first column
- No header cells

This field would be required. The main improvement here is the fact that contrary to checkboxes, the field has to be interacted with for the user to select the "No header cells" option that we want to discourage the use of (but should still be an option).

### Describe alternatives you've considered

One alternative would be to do this as a per-project customization, but it feels like an improvement that could easily live in Wagtail directly. Other changes to the TableBlock to remove the ability to have no header cells aren’t particularly compelling, because we still want the ability to select row/column/both.

### Additional context

Additionally to the above, as part of this change we could consider:

- Improving the help text for this field, to clarify its usage and the accessibility consequences
- Adding a straightforward way to customize this to disable the "No header cells" option – in practice all data tables should have header cells, and a table that doesn’t have any would be better as a list, or heading sections

In terms of migration, this looks like it could be easy:

- First row checkbox -> First row dropdown option
- First column checkbox -> First column dropdown option
- Both checkboxes -> "Both" dropdown option
- No checkboxes -> "No header cells" option

### PR DESCRIPTION
**Pull request by DevilsAutumn**
Fixes #5989 

Follow up to PRs  #9673 and #6763 

Before:
Screenshot 2023-12-09 at 4 36 02 PM
After:
Screenshot 2023-12-09 at 4 34 43 PM
Screenshot 2023-12-09 at 4 34 54 PM

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [x] **Please list the exact browser and operating system versions you tested**:
              os: macos sonoma 14.1.2
              browser: Firefox 120.0.1
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: client/src/entrypoints/contrib/table_block/table.js [hunk 1]
```diff
@@ -238,26 +233,23 @@ class TableInput {
...
+                ${this.strings['Display the first column as a header']}
+            </option>
+            <option value="both">
+                ${this.strings['Display the first row AND first column as headers']}
+            </option>
+            <option value="neither">
+                ${this.strings['No headers']}
+            </option>
+          </select>
+        <p class="help">${this.strings['Which cells should be displayed as headers?']}</p>
```
Comment (lb-) [inline client/src/entrypoints/contrib/table_block/table.js:252]: While we are here, can we associate this help text with the input - aria-describedby?

File: client/src/entrypoints/contrib/table_block/table.js [hunk 2]
```diff
@@ -123,8 +119,11 @@ function initTable(id, tableOptions) {
         data: hot.getData(),
         cell: cell,
         mergeCells: mergeCells,
-        first_row_is_table_header: tableHeaderCheckbox.prop('checked'),
-        first_col_is_header: colHeaderCheckbox.prop('checked'),
+        // first_row_is_table_header: tableHeaderCheckbox.prop('checked'),
+        // first_col_is_header: colHeaderCheckbox.prop('checked'),
```
Comment (lb-) [inline client/src/entrypoints/contrib/table_block/table.js:123]: should we remove these commented out lines?

**Review by lb- — COMMENTED** [review]
Just a few small comments, have not reviewed in full or run it locally.

Thanks for putting this PR together though! I will flag it for a full review.

File: client/src/entrypoints/contrib/table_block/table.js (hunk 1 above)
Comment (DevilsAutumn) [reply client/src/entrypoints/contrib/table_block/table.js:252]: im not good with the django templates yet but may i ask why with the input - aria-describedby? 🤔 
shouldn't the below changes be enough?
```suggestion
        <div class="w-field__help" aria-describedby="${id}-handsontable-col-caption-helptext" data-field-help>
            <p class="help">${this.strings['Which cells should be displayed as headers?']}</p>
          </div>
```

**Comment by cnk** [comment]
@DevilsAutumn Thanks for taking this on. I added some table blocks to my copy of BakeryDemo and the new select menu works well on tables created with it. 

Unfortunately it does not work if you already have tables with old style parameters. Do you think you could [have a look at Matt's comments](https://github.com/wagtail/wagtail/issues/5989#issuecomment-625497960) and work on automatically translating the checkbox-style data into the correct option for the select? When I use the new code to edit a page with a table with a first row header:

1.  the select menu is empty
2. I get the "preview out of data message" which means I have a validation error
3. The preview shows I have lost my first row header

Existing page:
Screenshot 2023-12-20 at 5 11 07 PM

What happens when I try to edit it: 
Screenshot 2023-12-20 at 5 10 37 PM

**Comment by cnk** [comment]
I took a stab add adding backwards compatibility. I am not sure if `to_python` is the correct place for this but it works.

**Comment by DevilsAutumn** [comment]
@cnk Thanks a lot for adding the backwards compatibility !

**Review by lb- — APPROVED** [review]
Looks good to me!

Thanks @DevilsAutumn  @cnk and @salty-ivy  for your original PRs and support with this.

I have tested an existing TableBlock a new one and all variations of the select field.

I will rebase, push and once CI runs will merge this in.

**Pull request merged**